Batched grad support for: slice, select, diagonal

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44505 Batched grad support for: slice, select, diagonal**
* #44494 Add some batched gradient tests

Added batching rules for slice_backward, select_backward, and
diagonal_backward.

Test Plan:
- new tests: `pytest test/test_vmap.y -v -k "BatchedGrad"`

Differential Revision: [D23650409](https://our.internmc.facebook.com/intern/diff/D23650409)

zou3519 committed:
Batched grad support for: slice, select, diagonal

Added batching rules for slice_backward, select_backward, and
diagonal_backward.

Test Plan:
- new tests: `pytest test/test_vmap.y -v -k "BatchedGrad"`

[ghstack-poisoned]

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -238,6 +258,15 @@ Tensor diagonal_batching_rule(const Tensor& self, int64_t offset, int64_t dim1,
 
+Tensor diagonal_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes, int64_t offset, int64_t dim1, int64_t dim2) {
+  auto grad_physical = MultiBatchVmapTransform::logicalToPhysical(grad);
+  auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
```
Do you really want to keep the grad's layout here? Or you want a contiguous layout (to make sure the .diagonal() call below won't fail)?

albanD replied:
Same for the other functions actually

zou3519 replied:
The options for memory format right now are all dense (contiguous memory format vs channels last memory format), so the diagonal() call will never fail. Also, if there is something wrong here, then we probably want to change https://github.com/pytorch/pytorch/blob/4bead6438a4271f154980dc242199c2f1b897bc1/torch/csrc/autograd/FunctionsManual.cpp#L1120-L1125 as well...

albanD replied:
Ok that sounds good.
I would guess that this is not tested though.
A simple thing like `a = torch.rand(10, 10, requires_grad=True); a.diagonal().sum().backward(); self.assertAllClose(a.grad, torch.eye(10))` could be added.
Do you think this can be added here or we should do that in a follow up PR?

zou3519 replied:
There are tests for diagonal_backward: https://github.com/pytorch/pytorch/blob/a2a81e1335f28dbd21ef517f18ff218706e70ffd/torch/testing/_internal/common_methods_invocations.py#L918-L929

Are you suggesting that we add a test for the case that `grad` is an expanded tensor? So something like:
```
a = torch.rand(10, 10, requires_grad=True)
constant = 3.14
v = torch.tensor([constant]).expand(10)
a_grad = torch.autograd.grad(a.diagonal(), a, v)
self.assertAllClose(a_grad, torch.eye(10) * constant)
```

albanD replied:
Yes something like that. That will not cover all the cases but would cover at least this one. You don't think that would be useful?

zou3519 replied:
I agree that that would be useful, I'll add it in this PR :)

zou3519 replied:
Hmm, but in general, it might be nice for gradcheck to pass in an expanded or non-contiguous `v` tensor to autograd.grad

albanD replied:
It's not really possible in general because we pass a very specific value (vector of zeros with a single one) and that cannot be done with an expanded Tensor.
But the gradgradcheck, you can generate non-contiguous grad_outputs. So this is tested for the functions where we do gradgradcheck.

zou3519 replied:
Updated with tests for torch.diagonal, torch.select, and torch.slice in the case where `v` is an expanded tensor

albanD commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -223,13 +223,33 @@ Tensor select_batching_rule(const Tensor& self, int64_t dim, int64_t index) {
+  auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
+  auto physical_dim = getGradInputPhysicalDim(dim, input_sizes, grad_physical.numBatchDims());
+  grad_input.select(physical_dim, index).copy_(grad_physical.tensor());
+  return grad_physical.newLogicalFromPhysical(grad_input);
```
I'm not sure to understand what this is doing. Where are these defined so that I can read about them?

zou3519 replied:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/VmapTransforms.h

Please feel free to ask questions about it, it is kinda confusing at times :/

zou3519 replied:
As a summary,
- auto grad_physical = MultiBatchVmapTransform::logicalToPhysical(grad);
    - This takes the logical BatchedTensor grad and turns it into a VmapPhysicalView. a VmapPhysicalView holds a regular (non-Batched) Tensor and some additional metadata. The regular (non-Batched) tensor is the grad tensor, but all of the batch/vmap dims are moved to the front of the tensor.
- auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
    - `grad_physical` is a VmapPhysicalView. Given the input_sizes, this prepends the batch (vmap) dimensions to the input_sizes.
- auto physical_dim = getGradInputPhysicalDim(dim, input_sizes, grad_physical.numBatchDims());
    - We want to translate "dim" to the correct physical dim on the grad_input Tensor. The grad_input Tensor has some number of batch dims at the front of the tensor and then some number of regular dims.
- grad_input.select(physical_dim, index).copy_(grad_physical.tensor());
    - grad_physical.tensor() pulls out the physical Tensor from the VmapPhysicalView
- return grad_physical.newLogicalFromPhysical(grad_input);
    - Now, grad_input is a regular Tensor. newLogicalFromPhysical takes `grad_input` and turns it into a BatchedTensor by marking the first `grad_physical.numBatchDim()` dimensions as batch/vmap dimensions.

zou3519 replied:
The note at https://github.com/pytorch/pytorch/blob/a2a81e1335f28dbd21ef517f18ff218706e70ffd/aten/src/ATen/BatchingRegistrations.cpp#L8-L49 should provide some context as well

ezyang replied:
Whether or not Alban can understand after these descriptions will be a good test, since I think this has been a me and you thing only so far XD

albanD replied:
I think this is quite good.
The main point of confusion for me was `newLogicalFromPhysical` that is called on a physical Tensor with a physical argument. So it is unclear from the name which one is used to create the logical output. Especially when you write `grad_physical.newLogicalFromPhysical(grad_input)` where the one with `_physical` in its name is not the one that is actually used.

I am not sure how to improve it though. It feels like there are two states in the naming right now: physical and logical. But the physical view should be there as well?
A random idea that comes to mind could be to have the 3 states be Logical, Physical and Plain? So that you have a `newLogicalFromPlain` function and `grad_physical.toPlain()` instead of `grad_physical.tensor()`.



The second point which is not really that important is that it's too bad that you cannot use `.getPhysicalShape()` here and have to re-implement it. You could maybe extend this API to have an overload that takes the plain sizes as input?

ezyang replied:
These are good points. One thing I'd point out is that we do encode Physical and Logical distinction in the type system, so the compiler will call you out if you get it wrong. But indeed some of these improvements may be worth doing.

zou3519 committed:
Update on "Batched grad support for: slice, select, diagonal"


Added batching rules for slice_backward, select_backward, and
diagonal_backward.

Test Plan:
- new tests: `pytest test/test_vmap.y -v -k "BatchedGrad"`

[ghstack-poisoned]

albanD reviewed (approved):
LGTM
The discussion about the vmap api is orthogonal to the PR content.
